### PR TITLE
[stable/redis] support redis docker library image

### PR DESCRIPTION
**Pull request by jlegrone**
**What this PR does / why we need it**:
~This release switches the default image from [`docker.io/bitnami/redis`](https://hub.docker.com/r/bitnami/redis/) to [`docker.io/redis`](https://hub.docker.com/r/library/redis/). The bitnami image remains compatible with this chart; see `values-bitnami.yaml`.~

The default image in this release may be switched out for any image containing the `redis-server` binary. If `redis-server` is not the default image ENTRYPOINT, `master.command` must be specified. 

This allows `stable/redis` to offer a higher degree of flexibility for those who may need to run images containing [redis modules](https://redis.io/modules) or based on a different linux distribution than what is currently offered by bitnami.

#### Breaking changes:
- `master.args` and `slave.args` are removed. Use `master.command` or `slave.command` instead in order to override the image entrypoint, or `master.extraFlags` to pass additional flags to `redis-server`.
- `disableCommands` is now interpreted as an array of strings instead of a string of comma separated values.
- `master.persistence.path` now defaults to `/data`.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*:

Related: https://github.com/bitnami/bitnami-docker-redis/issues/92

**Special notes for your reviewer**:

Several interesting test cases:
- `helm upgrade --install redis-test ./stable/redis -f ./stable/redis/values-bitnami.yaml`
- `helm upgrade --install redis-test ./stable/redis --set usePassword=false`
- `helm upgrade --install redis-test ./stable/redis --set master.persistence.enabled=false`
- `helm upgrade --install redis-test ./stable/redis --set configmap="maxmemory-policy volatile-lru"`
- `helm upgrade --install redis-test ./stable/redis --set image.repository=redislabs/rejson --set image.tag=latest`

### CODE REVIEW

**Comment by desaintmartin** [comment]
Thanks!
Let me just mention that we need to solve https://github.com/helm/charts/issues/7680#issuecomment-420693609 before any other contribution can be done on redis.
See https://github.com/helm/charts/pull/7745

File: stable/redis/values-production.yaml [hunk 1]
```diff
@@ -193,11 +189,12 @@ master:
   ## Update strategy, can be set to RollingUpdate or onDelete by default.
   ## https://kubernetes.io/docs/tutorials/stateful-application/basic-stateful-set/#updating-statefulsets
   statefulset:
-    updateStrategy: OnDelete
+    updateStrategy: RollingUpdate
```
Comment (prydonius) [inline stable/redis/values-production.yaml:214]: Why is this change needed? It is not described in the PR description.

**Review by prydonius — COMMENTED** [review]
This PR is doing two different things, that I think should be split into different PRs, and probably warrant some discussion.

1. Changing the default image to Docker Inc.'s Redis

I'm curious what the motivation for this change is -- is there something that Docker Inc.'s Redis image has that the Bitnami Redis image does not? In particular, changing the image of this chart has deeper consequences as it has automation setup to keep the bitnami/redis image up-to-date and secure (e.g. https://github.com/helm/charts/pull/7065).

2. Generalising the chart to support custom Redis images

I'm not fully convinced it makes sense for this chart to try to support every type of Redis image, they're all going to be a little bit different (env vars, volume paths, etc.). This sort of approach will make charts much more complex and harder to maintain and test. Instead, we can create different charts that are tailored to specific usecases (e.g. a redis chart that focuses on making it easy to add extra modules, or includes some more common modules vs a bare bones Redis chart). With the charts project moving to a more [distributed chart repositories model](https://github.com/helm/community/blob/master/proposals/distributed-search.md), it's going to be more common to see different variants of the same application for different usecases, and they can be hosted and maintained by anyone.

File: stable/redis/values-production.yaml (hunk 1 above)
Comment (jlegrone) [reply stable/redis/values-production.yaml:214]: I copied and pasted my updates from `values.yaml`, and I think there were a few additional things that had gotten out of sync. This didn't look like a value that should be different in prod environments.

**Comment by javsalgar** [comment]
> The problem is that I don't see a way to provide redis-server args directly to the bitnami/redis entrypoint. If this is already possible, or it can be supported in the future, I'm all for not switching out the default image. The bitnami image patch automation is great!

I think that, before chaning the image, we should see if we can implement the missing features in the Bitnami Redis image, so we do not lose all the security and automation features already available. I will forward this information for evaluation.

**Comment by juan131** [comment]
Hi @jlegrone 

I created this issue on Bitnami Redis Image to discuss how to implement the feature you're requesting:
https://github.com/bitnami/bitnami-docker-redis/issues/117

**Comment by juan131** [comment]
Hi @jlegrone 

We just released a new **Bitnami Redis image** including the changes that allow you to send the `redis-server` arguments directly to the entrypoint.

Please feel free to use it (**bitnami/redis:4.0.11-r50**) and provide us feedback about it. I think this image now supports the missing features so we can continue using **bitnami/redis** on the chart.

**Comment by jlegrone** [comment]
/remove-lifecycle stale

Sorry about the delay -- I made a big move recently and lost track of this PR. Will work on updating based on the recent work to modify the bitnami redis image.

**Comment by tuananh** [comment]
@juan131 hi, what's the recommended way to load an additional module with the chart?

**Comment by juan131** [comment]
Hi @tuananh 

The image doesn't support a way to do that via env. variables... I guess you can mount your module using a volume in the container and then user `redis-cli` to load it:

```
redis-cli -a YOURPASSWORD MODULE LOAD /path/to/the/module.so
```

You can open an issue at https://github.com/bitnami/bitnami-docker-redis so it's implemented a better way to do so.

**Comment by jlegrone** [comment]
Whew... finally getting back to this. This PR now does _not_ replace the bitnami image by default. 

Thanks @juan131 for upstreaming that change! Unfortunately there's one more thing that would be needed to get rid of the explicit `command` in `values.yaml`, but we can do that in a future patch if the entrypoint of the bitnami image does fully support redis args at some point. I'll file an issue to explain.

**Comment by tompizmor** [comment]
@jlegrone let us know when you create the issue and we will take a look.

File: stable/redis/README.md [hunk 2]
```diff
@@ -80,12 +91,12 @@ The following table lists the configurable parameters of the Redis chart and the
 |--------------------------------------------|----------------------------------------------------------------------------------------------------------------|------------------------------------------------------|
 | `global.imageRegistry`                     | Global Docker image registry                                                                                   | `nil`                                                |
 | `image.registry`                           | Redis Image registry                                                                                           | `docker.io`                                          |
-| `image.repository`                         | Redis Image name                                                                                               | `bitnami/redis`                                      |
+| `image.repository`                         | Redis Image name                                                                                               | `redis`                                              |
```
Comment (tompizmor) [inline stable/redis/README.md]: I guess this is `bitnami/redis` again no?

Comment (jlegrone) [reply stable/redis/README.md]: good catch, thanks

**Comment by tompizmor** [comment]
@jlegrone can you resolve the conflict in `stable/redis/Chart.yaml` ?

**Comment by tuananh** [comment]
@jlegrone nice <3

**Comment by qinix** [comment]
seems like `extraFlags` is not working 😟
